---
title: "Module_9_R"
output: html_document
---

### Variant statistics in R


#### Preparation
Load the required packages

```{r}
library(dplyr)
library(reshape2)
library(ggplot2)
```


Import the variant annotation data
```{r}

anno<-read.delim2("myanno.21.txt",header=F)
colnames(anno)<-c("chr","pos","raw_pos","ref","raw_ref","alt","raw_alt","mut_type","function","region_type","region_name","refGene.name","refGene.name2","dbSNP.name","ExAC_r0_3_sites_vep.name","ExAC_r0_3_sites_vep.AF","dbNSFP.SIFT_pred","dbNSFP.Polyphen2_HDIV_pred","dbNSFP.MutationTaster_pred","dbNSFP.MetaSVM_pred","dbNSFP.clinvar_clnsig","clinvar.CLNSIG")
```

clean up annot
```{r}
anno$chr<-paste0("chr",trimws(anno$chr))
anno$raw_pos<-trimws(anno$raw_pos)
anno$raw_ref<-trimws(anno$raw_ref)
anno$raw_alt<-trimws(anno$raw_alt)
```

create a unique variant ID
```{r}
anno$varID<-paste(anno$chr,anno$raw_pos,anno$raw_ref,anno$raw_alt,sep=":")
```


Import genotypes
```{r}
geno<-read.delim2("GMTS.geno.txt")

colnames(geno)<-c("chr","raw_pos","id","raw_ref","raw_alt","qual","filter","AC","AF","AN","BaseQRankSum","ClippingRankSum","DP","DS","END","ExcessHet","FS","Hrun","HaplotypeScore","InbreedingCoeff","MLEAC","MLEAF","MQ","MQRankSum","QD","RAW_MQ","ReadPosRankSum","SOR","SAMPLE","AD","DP.1","GQ","GT","MIN_DP","PGT","PID","PL","RGQ","SB")
```


create a unique variant ID
```{r}
geno$varID<-paste(geno$chr,geno$raw_pos,geno$raw_ref,geno$raw_alt,sep=":")
```

add indel info
```{r}
geno<-geno %>% mutate(type=ifelse(nchar(as.character(raw_ref))>1 | nchar(as.character(raw_alt))>1,"indel","SNV"))
```

add sample IDs. 
```{r}
idxref<-read.delim2("idxref.txt",sep=" ")
colnames(idxref)<-c("sampleID","SAMPLE")

geno<-left_join(geno,idxref)
```

import metadata
```{r}
meta<-read.delim2("GMTS_metadata.txt")
```



#### Quality controls


extract key quality metrics
```{r}
qual<-geno %>% select(varID,type,DP,FS,Hrun,MQ,QD)
```

Depth distribution
```{r}
ggplot(qual,aes(log10(as.numeric(DP)),col=type))+geom_density()
```


FisherStrand Bias distribution
```{r}
ggplot(qual,aes(as.numeric(FS),col=type))+geom_density()
```

count genoypes by GT categories
```{r}
geno %>% group_by(GT,type) %>% tally()
```

keep good genotype with low strand bias

```{r}
geno<-geno %>% filter(GT=="0/0" | GT=="0/1" | GT=="1/1") %>% filter(as.numeric(FS)<60) %>% filter(as.numeric(QD)>2)
```


remove all the variants no longer variants
```{r}
to_keep<-geno %>% filter(GT=="0/1" | GT=="1/1") %>% select(varID)
geno<-geno %>% filter(varID %in% to_keep$varID)
```

recounting
```{r}
geno %>% group_by(GT,type) %>% tally()
```



#### combine

Join genotypes and annotations
```{r}

```

Join with the phenotypes
```{r}

```


Flag the deleterious variants
```{r}

```

Compare the genetic burden for each gene between male and female
```{r}

```

Plot the resultign volcano plot



```{r}

```